File: mcp/src/selftest.ts
/**
 * Runs the shared fixture file. Same cases, same expected verdicts as Python.
 * A disagreement here means the two implementations have drifted.
 */

import { existsSync, readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

import {
  checkKeywords, checkTranslation, checkInternalLinks, checkClaims,
  contentHash, resolveLastmod, DEFAULT_PRIMARY_MIN_HITS,
} from "./guards.js";

const here = dirname(fileURLToPath(import.meta.url));

// Inside the published package the fixtures sit next to dist/, copied there by
// the prebuild step. In the repo they live at the top level, shared with the
// Python implementation. A published package that cannot find its own tests
// would refuse to start, so both locations are tried.
const CANDIDATES = [
  resolve(here, "..", "fixtures", "cases.json"),
  resolve(here, "..", "..", "fixtures", "cases.json"),
];

export const FIXTURES = CANDIDATES.find(existsSync) ?? CANDIDATES[0];

type Verdict = "pass" | "block";

interface Case { guard: string; name: string; expect: Verdict; input: Record<string, any>; }

export function verdict(c: Case): Verdict {
  const i = c.input;
  switch (c.guard) {
    case "keywords": {
      const { problems } = checkKeywords(
        i.html, i.keywords, i.scope ?? null, i.exclude_classes ?? [],
        i.primary_min_hits ?? DEFAULT_PRIMARY_MIN_HITS);
      return problems.length ? "block" : "pass";
    }
    case "translation": {
      const problems = checkTranslation(i.source, i.translation, i.locale, {
        forbidSourceScript: i.forbid_source_script ?? true,
        checkLocalePrefix: i.check_locale_prefix ?? true,
      });
      return problems.length ? "block" : "pass";
    }
    case "internal_links": {
      const { problems } = checkInternalLinks(
        i.html, i.money_paths, i.scope ?? null, i.exclude_classes ?? [],
        i.require_figure ?? false);
      return problems.length ? "block" : "pass";
    }
    case "claims":
      return checkClaims(i.text).length ? "block" : "pass";
    case "lastmod": {
      const strip = i.strip_classes ?? [];
      const prev = contentHash(i.previous_body, strip);
      const got = resolveLastmod(i.body, prev, i.previous_lastmod, i.today, strip);
      return got.lastmod === i.expect_lastmod ? "pass" : "block";
    }
    default:
      throw new Error("unknown guard in fixtures: " + c.guard);
  }
}

export function run(log: (s: string) => void = console.log): boolean {
  const data = JSON.parse(readFileSync(FIXTURES, "utf8")) as { cases: Case[] };
  let passed = 0, failed = 0, current = "";
  for (const c of data.cases) {
    if (c.guard !== current) { current = c.guard; log("  " + current); }
    const got = verdict(c);
    const ok = got === c.expect;
    ok ? passed++ : failed++;
    log(`    [${ok ? "ok" : "FAIL"}] expected ${c.expect.padEnd(5)} got ${got.padEnd(5)}  ${c.name}`);
  }
  log(`  ${passed} passed, ${failed} failed`);
  return failed === 0;
}

if (process.argv[1] && process.argv[1].endsWith("selftest.js")) {
  process.exit(run() ? 0 : 2);
}

File: src/contentgate/selftest.py
"""Run the shared negative-test cases against this implementation.

A guard that has never been shown to fail is decoration. Every case in
fixtures/cases.json states an expected verdict, and half of them expect the
guard to block. The TypeScript implementation runs the same file, so the two
cannot quietly drift apart.
"""

import io
import os
import json
import pathlib

from .guards import keywords, translation, internal_links, claims, lastmod

FIXTURES = pathlib.Path(__file__).resolve().parents[2] / "fixtures" / "cases.json"


def _verdict(case):
    """Return 'pass' or 'block' for one case."""
    g, i = case["guard"], case["input"]

    if g == "keywords":
        problems, _ = keywords.check(
            i["html"], i["keywords"],
            scope=tuple(i["scope"]) if i.get("scope") else None,
            exclude_classes=i.get("exclude_classes", ()),
            primary_min_hits=i.get("primary_min_hits", keywords.DEFAULT_PRIMARY_MIN_HITS))
        return "block" if problems else "pass"

    if g == "translation":
        problems = translation.check(
            i["source"], i["translation"], i["locale"],
            forbid_source_script=i.get("forbid_source_script", True),
            check_locale_prefix=i.get("check_locale_prefix", True))
        return "block" if problems else "pass"

    if g == "internal_links":
        problems, _ = internal_links.check(
            i["html"], i["money_paths"],
            scope=tuple(i["scope"]) if i.get("scope") else None,
            exclude_classes=i.get("exclude_classes", ()),
            require_figure=i.get("require_figure", False))
        return "block" if problems else "pass"

    if g == "claims":
        return "block" if claims.check(i["text"]) else "pass"

    if g == "lastmod":
        strip = i.get("strip_classes", ())
        prev_hash = lastmod.content_hash(i["previous_body"], strip)
        got, _changed, _h = lastmod.resolve(
            i["body"], prev_hash, i["previous_lastmod"], i["today"], strip)
        return "pass" if got == i["expect_lastmod"] else "block"

    raise ValueError("unknown guard in fixtures: " + g)


def run(stream=None, fixtures=FIXTURES):
    """Execute every case. Returns (ok, passed, failed)."""
    out = stream if stream is not None else io.StringIO()
    data = json.loads(pathlib.Path(fixtures).read_text(encoding="utf-8"))
    passed = failed = 0
    current = None
    for case in data["cases"]:
        if case["guard"] != current:
            current = case["guard"]
            print("  " + current, file=out)
        got = _verdict(case)
        ok = got == case["expect"]
        passed, failed = (passed + 1, failed) if ok else (passed, failed + 1)
        print("    [{}] expected {:<5} got {:<5}  {}".format(
            "ok" if ok else "FAIL", case["expect"], got, case["name"]), file=out)
    print("  {} passed, {} failed".format(passed, failed), file=out)
    return failed == 0, passed, failed


if __name__ == "__main__":
    import sys
    ok, p, f = run(sys.stdout)
    sys.exit(0 if ok else 2)
